### ISSUE TITLE
OverflowException in SCardReaderState.get_Atr()

### ISSUE THREAD

**Issue by JssDWt** [danm-de/pcsc-sharp#87]
Probably another int32/int64 problem here:

I'm running the 'MonitorReaderEvents' sample on a MacBook Pro with a SCL3711 attached in a .Net Core 2.1 app. When I get my phone in range, however, I'm getting an OverFlowException:

> This program will monitor all SmartCard readers and display all status changes.
> Start monitoring for reader SCL3711 Reader and NFC Device.
> Press Ctrl-Q to exit or any key to toggle monitor.
> >> Initialized Event for reader: SCL3711 Reader and NFC Device
> ATR: 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00
> State: Unaware
> 
> 
> Unhandled Exception: System.OverflowException: Arithmetic operation resulted in an overflow.
>    at PCSC.SCardReaderState.get_Atr()
>    at PCSC.Monitoring.SCardMonitor.StartMonitor(Monitor monitor)
>    at System.Threading.Thread.ThreadMain_ParameterizedThreadStart(Object parameter)
>    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()

**Comment by JssDWt** [comment]
I thought the code should become something like this:
```csharp
public byte[] Atr {
            get {
                byte[] tmp;

                if (Platform.IsWindows) {
                    if (_winscardRstate.cbAtr <= WinSCardAPI.MAX_ATR_SIZE) {
                        tmp = new byte[_winscardRstate.cbAtr];
                    } else {
                        // error occurred during SCardGetStatusChange()
                        tmp = new byte[WinSCardAPI.MAX_ATR_SIZE];
                        _winscardRstate.cbAtr = WinSCardAPI.MAX_ATR_SIZE;
                    }

                    Array.Copy(_winscardRstate.rgbAtr, tmp, _winscardRstate.cbAtr);
                } else {
                    var cbAtr = _pcscliteRstate.cbAtr.ToInt64();
                    if (cbAtr <= PCSCliteAPI.MAX_ATR_SIZE) {
                        tmp = new byte[cbAtr];
                    } else {
                        // error occurred during SCardGetStatusChange()
                        tmp = new byte[PCSCliteAPI.MAX_ATR_SIZE];
                        _pcscliteRstate.cbAtr = (IntPtr) PCSCliteAPI.MAX_ATR_SIZE;
                    }

                    Array.Copy(_pcscliteRstate.rgbAtr, tmp, tmp.Length);
                }

                return tmp;
            }
```

However, When I run the program now, and move my phone over the reader, I get status 'Ignore, Unknown', or 'Ignore, Unkown, Unavailable' While I think this should be status 'Present' or 'Present, InUse'. Also, the state does not change back when I remove the phone. So perhaps something else is wrong.

> This program will monitor all SmartCard readers and display all status changes.
> Start monitoring for reader SCL3711 Reader and NFC Device.
> Press Ctrl-Q to exit or any key to toggle monitor.
> >> Initialized Event for reader: SCL3711 Reader and NFC Device
> ATR: 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00
> State: Unaware
> 
> >> StatusChanged Event for reader: SCL3711 Reader and NFC Device
> ATR: 00
> Last state: Unaware
> New state: Ignore, Unknown

**Comment by danielmue** [comment]
The MacOS support is new: https://github.com/danm-de/pcsc-sharp/pull/85 (contributed by @pedromrpinto)

Could you write me the required software / steps, needed to reproduce the problem, please?

**Comment by JssDWt** [comment]
## Steps to reproduce
- Install [mac driver for SCL3711](http://files.identiv.com/products/smart-card-readers/contactless/scl3711/SCx371x_MacOS_Driver.zip) (Guess this would be the same for other cards)
- Plug in SCL3711 via USB
- Clone this repo
- Create a new .Net core console application, reference the PCSC project, and copy the contents of Examples/MonitorReaderEvents/Program.cs to the newly created Console application.
- Run the new console application. (Make sure it starts in an external terminal)
- You will get output as in [this comment](https://github.com/danm-de/pcsc-sharp/issues/87#issue-411789991)
- Change the code of SCardReaderState.get_Atr() as shown in [this comment](https://github.com/danm-de/pcsc-sharp/issues/87#issuecomment-465111793) 
- Run the app again.
- Hold an NFC enabled smartphone over the reader. Output will be the bottom of [this comment](https://github.com/danm-de/pcsc-sharp/issues/87#issuecomment-465111793)

## Notes:
- Running a sample app from the Springcard sdk does return status 'Present' or 'Present, InUse'.
- The actual status returned by the internal Unix code is 108227381948317701

**Comment by danm-de** [comment]
Okay, thanks. I'll try to organize a MacBook. Seems that libpcsclite behaves differently on MacOS. I hope the method signatures (c header files) are the same - the separation between Windows & UNIX (Linux) Interop has been made for a reason :-( 
That "int vs. long" dilemma bugs me since the first release.

Your received state 108227381948317701(10) = 180803B00000005(16) is odd. Looks pretty much like  wrong data type usage during P/Invoke call...

**Comment by melihercan** [comment]
Hi folks,
I am developing a Xamarin Forms app with MacOS support and suffering from the same issue. The reader is  Idendiv uTrust 3700 F. I attach the screen shot hoping that it might help fixing the issue.
<img width="1794" alt="MacOS_Exception" src="https://user-images.githubusercontent.com/10947383/54937633-1539bb00-4f25-11e9-91d3-1b51de00acea.png">

**Comment by melihercan** [comment]
I am back to this and it seems the issue is in the definition of SCARD_READERSTATE struct in Unix/ExtensionMethods/SCARD_READERSTATE.cs file:
internal struct SCARD_READERSTATE
    {
        internal IntPtr pszReader;
        internal IntPtr pvUserData;
        internal IntPtr dwCurrentState;
        internal IntPtr dwEventState;
        internal IntPtr cbAtr;

        [MarshalAs(UnmanagedType.ByValArray, SizeConst = PCSCliteAPI.MAX_ATR_SIZE)]
        internal byte[] rgbAtr;
    }
Here dwCurrentState, dwEventState and cbAtr are defined as DWORD in native struct just like in Windows case. In Windows these prameters are defined as "int", NOT as "IntPtr". I changed them to "int" and adapted the related get/set operations in SCardReaderState.cs then it started working.

**Comment by danm-de** [comment]
Hi @melihercan,
thanks for reporting. I guess you are a Mac owner -> Could you please checkout this branch?
-> https://github.com/danm-de/pcsc-sharp/tree/bugfix/macos-issue-87

Does this work for you?

**Comment by melihercan** [comment]
Hi Daniel,

I will test it and get back to you. It may take few days as I am very busy
right now.

Thanks,
Melih

**Comment by melihercan** [comment]
Hi again Daniel,
Finally I had some time to check the branch. Well good news, it seems it is working now. I am able to read the tag.

**Comment by danm-de** [comment]
Hi Melih,
thanks for your feedback! A friend of mine tried it as well - using the latest mono runtime. The monitor example throws segmentation faults and PlatformNotSupportedExceptions. 

I am not sure what's wrong here. 
- Unsupported PC/SC driver?
- Wrong P/Invoke call (maybe another 32/64 bit data type issue)?
Since SCARD_READERSTATE differs from the Linux version I can imagine the same applies to other API signatures.

**Comment by danm-de** [comment]
I have got a MacBook to test. After checking pcsclite.h, wintypes.h and winscard.h [0] I can confirm that **all** P/Invoke method signatures and data structures are wrong for MacOS X. :-(

I've started to implement the changes in the same branch. 

[0] /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/PCSC.framework/Versions/Current/Headers

**Comment by danm-de** [comment]
Should be fixed in version 4.1.1.
